[cxx-interop] Instantiate C++ class templates from Swift

hlopko opened the pull request:
This PR makes it possible to instantiate C++ class templates from Swift. Given a C++ header:

```c++
// C++ module `ClassTemplates`
template<class T>
struct MagicWrapper {
  T t;
};

struct MagicNumber {};
```

it is now possible to write in Swift:

```swift
import ClassTemplates

func x() -> MagicWrapper<MagicNumber> {
  return MagicWrapper<MagicNumber>()
}
```

This is achieved by importing C++ class templates as generic structs, and then when Swift type checker calls `applyGenericArguments` we detect when the generic struct is backed by the C++ class template and call Clang to instantiate the template. In order to make it possible to put class instantiations such as `MagicWrapper<MagicNumber>` into Swift signatures, we have created a new field in `StructDecl` named `TemplateInstantiationType` where the typechecker stores the `BoundGenericType` which we serialize. Deserializer then notices that the `BoundGenericType` is actually a C++ class template and performs the instantiation logic.

Depends on https://github.com/apple/swift/pull/33420.
Progress towards https://bugs.swift.org/browse/SR-13261.
Fixes https://bugs.swift.org/browse/SR-13775.

hlopko commented:
This PR depends on https://github.com/apple/swift/pull/32950.

hlopko commented:
@swift-ci please test

hlopko commented:
@swift-ci please test

hlopko commented:
@swift-ci please test

hlopko commented:
@swift-ci please test

hlopko commented:
@swift-ci please test

hlopko commented:
@swift-ci please test

hlopko commented:
@swift-ci please test

hlopko commented:
@swift-ci please test

hlopko commented:
@swift-ci please test

hlopko commented:
@swift-ci please test

zoecarver commented on lib/Serialization/ModuleFormat.h:
```diff
@@ -1703,6 +1703,12 @@ namespace decls_block {
     BCVBR<5>  // index
   >;
 
+  using XRefClangTemplateInstantiationLayout = BCRecordLayout<
```
Do we need to bump `SWIFTMODULE_VERSION_MINOR`?

hlopko replied:
Done

zoecarver commented on test/Interop/Cxx/templates/Inputs/class-template-non-type-parameter.h:
```diff
@@ -0,0 +1,11 @@
+    T t[Size];
+};
+
+typedef MagicArray<int, 2> MagicIntPair;
```
What about a failing test to make sure we don't crash if the user tries to instantiate `MagicArray` or a similar type with a non-type template parameter that's defaulted?

hlopko replied:
Great point! :) So I made MagicArray parameters defaulted and all works fine when they're explicitly mentioned, but not when we want to rely on the default. This could be improved, but I'd leave it for a separate PR as I expect quite a bit discussion over that.

zoecarver replied:
Sounds good to me.

zoecarver commented on test/Interop/Cxx/templates/class-template-instantiation-error-handling.swift:
```diff
@@ -0,0 +1,19 @@
+
+// CHECK: class-template-instantiation-error-handling.swift:10:18: error: only C++ types supported
+TemplatesTestSuite.test("swift-template-arg-not-supported") {
+  var magicInt = MagicWrapper<String>(t: "asdf")
```
nit: `magicInt` -> `magicString`

hlopko replied:
Done.

zoecarver commented on test/Interop/Cxx/templates/class-template-variadic.swift:
```diff
@@ -0,0 +1,27 @@
+
+var TemplatesTestSuite = TestSuite("TemplatesTestSuite")
+
+TemplatesTestSuite.test("variadic-class-template") {
```
Could you add a test for `Tuple` itself (not just `Pair`)?

hlopko replied:
Added a failing test and a TODO. Swift currently sees it as if we passed too many template arguments.

varungandhi-apple commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -2758,6 +2758,16 @@ ERROR(unexportable_clang_function_type,none,
+      "only struct/class/enum/union type arguments supported",
+      ())
+ERROR(cxx_class_instantiation_non_cxx_argument,none,
+      "only C++ types supported",
```
Could you add tests for these new diagnostics?

varungandhi-apple replied:
Also, for now, diagnostics need to be copied to localization/diagnostics/en.yaml as well. https://github.com/apple/swift/pull/32984#discussion_r456848980

hlopko replied:
Two of these diagnostics were already tested in`test/Interop/Cxx/templates/class-template-instantiation-error-handling.swift`. The third one I couldn't come up with a test, so I converted the diagnostic with an assertion.

zoecarver commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -830,6 +833,41 @@ static Type applyGenericArguments(Type type, TypeResolution resolution,
+  if (auto clangDecl = decl->getClangDecl()) {
+    if (auto classTemplateDecl =
+            dyn_cast<clang::ClassTemplateDecl>(clangDecl)) {
+      SmallVector<clang::TemplateArgument, 2> templateArguments;
```
When you want to support non-clang types, it shouldn't be too hard to replace this with `ASTContext ::getClangTemplateArguments`.

zoecarver replied:
Then we'll have the same issue with Swift generic types, though. Anyway, that's a problem for later 😅

hlopko replied:
Yup I'll definitely use `getClangTemplateArguments` once it lands.

zoecarver replied:
`getClangTemplateArguments` is implemented in master, if you want to use it in this patch. Otherwise, it could certainly be a follow up.

hlopko replied:
Done, thanks for reminding me.

varungandhi-apple commented:
Could you please rewrite the history into self-contained commits with commit messages that describe what the change in each commit is?

hlopko committed:
This change makes it possible to instantiate C++ class templates from Swift.

Given a C++ header:

```c++
// C++ module `ClassTemplates`
template<class T>
struct MagicWrapper {
  T t;
};

struct MagicNumber {};
```

it is now possible to write in Swift:

```c++
import ClassTemplates

func x() -> MagicWrapper<MagicNumber> {
  return MagicWrapper<MagicNumber>()
}
```

This is achieved by importing C++ class templates as generic structs,
and then when Swift type checker calls applyGenericArguments we detect
when the generic struct is backed by the C++ class template and call
Clang to instantiate the template. In order to make it possible to put
class instantiations such as MagicWrapper<MagicNumber> into Swift
signatures, we have created a new XREF
(XREF_CLANG_TEMPLATE_INSTANTIATION) into serialisation and
deserialisation infrastructure.

Depends on #33420.
Progress towards https://bugs.swift.org/browse/SR-13261.

hlopko commented:
> Could you please rewrite the history into self-contained commits with commit messages that describe what the change in each commit is?

I generally `squash and merge` when merging the PR, so at the end there's just one commit with the description being the description of the PR. This PR got out of hand with commits though so I'll squash them now too :)

hlopko committed:
Bump SWIFTMODULE_VERSION

hlopko committed:
Remove unreachable diagnostics

hlopko committed:
Rename instantiateTemplates

hlopko committed:
Fix tests

hlopko committed:
Add failing test for defaulted params

hlopko committed:
Add more failing tests

hlopko commented on include/swift/AST/ClangModuleLoader.h:
```diff
@@ -174,6 +175,14 @@ class ClangModuleLoader : public ModuleLoader {
                       llvm::function_ref<void(TypeDecl *)> receiver) = 0;
+  /// Instantiate and import class template.
+  virtual NominalTypeDecl *
+  instantiateTemplate(clang::ClassTemplateDecl *decl,
```
Done, went with `instantiateCXXClassTemplate`.

hlopko commented:
Thanks so much for the first round of comments!

hlopko commented:
@swift-ci please test

hlopko commented:
@swift-ci please test windows platform

varungandhi-apple commented on test/Interop/Cxx/templates/class-template-instantiation-error-handling.swift:
```diff
@@ -0,0 +1,19 @@
+// RUN: not %target-swift-emit-sil %s -I %S/Inputs -enable-cxx-interop 2>&1 | %FileCheck %s
```
Could you use `%target-swift-emit-sil -verify` (I think that should work...) instead of using `%FileCheck` here? We generally check diagnostics using `-verify`, not with `FileCheck`. If the errors are being emitting during SILGen, you could use `%target-swift-emit-silgen`.

hlopko replied:
Is there a way to expect an error happening in a different file? `-verify` seems to assume the errors to happen in the current file, but in my case I want assert that the error shows location in the header where the instantiation happened.

varungandhi-apple replied:
(Sorry, missed this comment earlier.)

> in my case I want assert that the error shows location in the header where the instantiation happened.

I see, I didn't realize that was happening. That makes sense.

varungandhi-apple commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -2743,6 +2743,36 @@ void ClangImporter::lookupRelatedEntity(
+  }
+
+  auto CanonType = decl->getASTContext().getTypeDeclType(ctsd);
+  assert(isa<clang::RecordType>(CanonType) &&
+          "type of non-dependent specialization is not a RecordType");
```
Could this happen because of an internal error (this check is already handled elsewhere but maybe there's a bug there) or is this potentially because of a user error? If it could be because of a user error, please at least add a TODO for surfacing a diagnostic here.

hlopko replied:
I'm closely following the logic in Clang (https://github.com/llvm/llvm-project/blob/1abb883a048153c83a4e11070219d23f362e7377/clang/lib/Sema/SemaTemplate.cpp#L3651), so it should be a programmer error when this hits, not a user error. I'm happy to remove the assertion if you think it doesn't make sense to have it.

varungandhi-apple replied:
(Sorry, missed this earlier.)

If Clang also does the same thing, I think it's fine. 👍🏼

varungandhi-apple commented on test/Interop/Cxx/templates/swift-class-instantiation-module-interface.swift:
```diff
@@ -0,0 +1,10 @@
+// RUN: %target-swift-ide-test -print-module -module-to-print=SwiftClassInstantiationModule -I %t -I %S/Inputs -source-filename=x -enable-cxx-interop | %FileCheck %s
+
+// CHECK: import MagicWrapper
+// CHECK: import SwiftOnoneSupport
```
Is there a specific reason we are checking for `import SwiftOnoneSupport` here? We generally emit it for (almost?) every module, so I don't think we need to check for it explicitly here.

hlopko replied:
No good reason, I just copy-pasted the entire output of `swift-ide-test`. Cleaned up now.

hlopko committed:
Cleanup module interface test

hlopko commented on test/Interop/Cxx/templates/swift-class-instantiation-module-interface.swift:
```diff
@@ -0,0 +1,10 @@
+// RUN: %empty-directory(%t)
+// RUN: %target-build-swift %S/Inputs/SwiftClassInstantiationModule.swift -module-name SwiftClassInstantiationModule -emit-module -emit-module-path %t/ -I %S/Inputs -Xfrontend -enable-cxx-interop -parse-as-library
+// RUN: %target-swift-ide-test -print-module -module-to-print=SwiftClassInstantiationModule -I %t -I %S/Inputs -source-filename=x -enable-cxx-interop | %FileCheck %s
```
Done, thank you!

Regarding mangled names - those are just names of structs generated for template instantiation. We needed a way to generate a name that encodes the class template name plus all its arguments, and that's non-trivial for C++ types. Therefore we took itanium mangled name and prepended `__CxxTemplateInst`. This naming was estabilished in https://github.com/apple/swift/pull/33451.

hlopko commented:
@swift-ci please test

zoecarver reviewed (approved):
This LGTM, thanks!

zoecarver commented:
Here's a test case that crashes:
```
template<class...>
class Foo;

template<>
class Foo<> {
  enum Maybe { No, Yes };
};
```
But, I think it would be fair to fix that in another patch.

hlopko committed:
Merge branch 'main' into class_templates

hlopko commented:
@swift-ci please test

hlopko committed:
Fixes

zoecarver commented on test/Interop/Cxx/templates/Inputs/class-template-instantiation-errors.h:
```diff
@@ -0,0 +1,22 @@
+#ifndef TEST_INTEROP_CXX_TEMPLATES_INPUTS_CLASS_TEMPLATE_INSTANTIATION_ERRORS_H
+#define TEST_INTEROP_CXX_TEMPLATES_INPUTS_CLASS_TEMPLATE_INSTANTIATION_ERRORS_H
+
```
Would you mind adding the following:
```
template<class T>
 struct MagicWrapper {
   T t;
   int getValuePlusArg(T arg) const { return t.getValue() + arg.getValue(); }
 };
```

zoecarver replied:
Maybe also a nested enum or union type?

hlopko replied:
I'm sorry, what exactly do you want to cover with that test? The difference between that and the template at the beginning of this file is that in your version we call `arg.getValue()`, right? Could you also elaborate on why nested enum/union types should be tested, what failure case do you envision?

zoecarver replied:
Not sure, but I'm worried a member that has a template argument might crash when getting imported as a generic class. When the function decl visitor goes looking for where the template param type it likely won't be able to find it. We should make sure we at least don't import that member (and make sure we don't crash).

As for nested enum/union types, I don't think that's as important to cover, but it is a case we should make sure we don't crash on that I don't think we currently have coverage for. Here's an example:
```
template<class T>
struct Parent {
  enum Foo : T { A, B, C };
};
```
Maybe just enums (not unions after all).

hlopko replied:
Hm, I tried this and it seems we import Foo as a struct. I don't yet understand why, but I guess this can be addressed in a separate PR, as it's also related to typedeffed templates. SG?

hlopko committed:
Use getClangTemplateArguments

hlopko commented:
@swift-ci please test

hlopko committed:
Merge branch 'main' into class_templates

hlopko commented:
@swift-ci please test

hlopko committed:
Fix symbol visitibity test

hlopko committed:
Merge branch 'main' into class_templates

zoecarver commented on test/Interop/Cxx/templates/class-template-instantiation-in-swift.swift:
```diff
@@ -0,0 +1,16 @@
+  var intWrapper = IntWrapper(value: 42)
+  var wrappedMagicNumber = MagicWrapper<IntWrapper>(t: intWrapper)
+  expectEqual(wrappedMagicNumber.getValuePlusArg(8), 50)
+}
```
Would you mind adding another test for `MagicWrapper<Int>` (or some test where T = a swift type)?

hlopko replied:
Done. I added a test using CInt as a param. And in error reporting test we have a test showing the error message when importing a Swift type that cannot be (currently) converted - Optional. 

I just discovered that this PR fixes https://bugs.swift.org/browse/SR-13775, so I uncommented the test and removed the early bail-out added by https://github.com/apple/swift/pull/34428. \o/

zoecarver replied:
Great! Thanks.

hlopko committed:
Dont use String because on mac it will turn to be NSString

hlopko commented:
@swift-ci please test

hlopko committed:
Add CInt test

hlopko committed:
Add test

hlopko commented:
@swift-ci please test.

Commit:
Update include/swift/AST/DiagnosticsSema.def

Co-authored-by: Dmitri Gribenko <gribozavr@gmail.com>

Commit:
Update lib/ClangImporter/ClangImporter.cpp

Co-authored-by: Dmitri Gribenko <gribozavr@gmail.com>

hlopko committed:
Address comments

hlopko commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -3988,6 +3981,10 @@ namespace {
         markAsVariant(result, *correctSwiftName);
 
+      if (decl->isInvalidDecl()) {
+        Impl.markUnavailable(result, "function marked as invalid by clang");
```
Removed, the only reason I wrote it was consistency with clang logic.

hlopko commented on test/stdlib/symbol-visibility-linux.test-sh:
```diff
@@ -35,6 +35,8 @@
 // RUN:             -e _ZSt16__once_call_implISt12_Bind_simpleIFPFvPvEPA216_cEEEvv \
 // RUN:             -e _ZN9__gnu_cxx12__to_xstringISscEET_PFiPT0_mPKS2_P13__va_list_tagEmS5_z \
+// RUN:             -e _ZZNSt8__detail18__to_chars_10_implIjEEvPcjT_E8__digits \
+// RUN:             -e _ZZNSt8__detail18__to_chars_10_implImEEvPcjT_E8__digits \
```
Will upload as a separate PR.

hlopko commented on test/Interop/Cxx/templates/Inputs/magic-wrapper.h:
```diff
@@ -0,0 +1,22 @@
+};
+
+template<class T>
+struct MagicWrapper {
```
Forked into multiple standalone headers, thanks!

hlopko commented on test/Interop/Cxx/templates/class-template-non-type-parameter.swift:
```diff
@@ -3,13 +3,29 @@
+// TODO(SR-13261): This test doesn't work because Swift only expects types as generic arguments.
+// TemplatesTestSuite.test("non-type-parameter") {
+//   var intWrapper = IntWrapper(value: 5)
+//   var pair = MagicArray<Element, intWrapper>(data: (Element(id: 1), Element(id: 2), Element(id: 3), Element(id: 4), Element(id: 5)))
```
Cleaned up the code, got out of sync a bit since commented out

hlopko commented on test/Interop/Cxx/templates/class-template-instantiation-error-handling.swift:
```diff
@@ -0,0 +1,19 @@
+// RUN: not %target-swift-emit-sil %s -I %S/Inputs -enable-cxx-interop 2>&1 | %FileCheck %s
```
Done.

hlopko commented on test/Interop/Cxx/templates/class-template-instantiation-error-handling.swift:
```diff
@@ -0,0 +1,19 @@
+    var _ = CannotBeInstantianted<IntWrapper>()
+}
+
+runAllTests()
```
Done

hlopko commented on test/Interop/Cxx/templates/class-template-instantiation-demangling.swift:
```diff
@@ -0,0 +1,12 @@
+// RUN: %target-swift-emit-ir %s -I %S/Inputs -enable-cxx-interop | grep 'define.*swiftcc.*$' | grep -o '[[:alnum:]]*__CxxTemplateInst[[:alnum:]]*' | xargs %swift-demangle | %FileCheck %s
```
Nope, it's the same, probably cp instead of mv :)

hlopko commented on test/Interop/Cxx/templates/not-pre-defined-class-template.swift:
```diff
@@ -7,10 +7,16 @@ import StdlibUnittest
   expectEqual(magicInt.getValuePlusArg(11), 28)
 }
 
+TemplatesTestSuite.test("existing-specialization") {
```
I extracted a standalone test, thanks for noticing!

hlopko commented:
@swift-ci please test

hlopko commented:
Addressed the comments, ptal :)

gribozavr commented on lib/Serialization/ModuleFormat.h:
```diff
@@ -1703,6 +1703,12 @@ namespace decls_block {
 
+  using XRefClangTemplateInstantiationLayout = BCRecordLayout<
+    XREF_CLANG_TEMPLATE_INSTANTIATION,
+    IdentifierIDField, // template name
```
The template may be in a namespace.

gribozavr commented on lib/Serialization/ModuleFormat.h:
```diff
@@ -1703,6 +1703,12 @@ namespace decls_block {
+  using XRefClangTemplateInstantiationLayout = BCRecordLayout<
+    XREF_CLANG_TEMPLATE_INSTANTIATION,
+    IdentifierIDField, // template name
+    BCArray<IdentifierIDField> // template args
```
Template args are not necessarily single identifiers -- the types may be nested in namespaces, may be template specializations themselves, and even can be non-type template parameters.

Is there an existing way to serialize a Clang type?

hlopko committed:
Merge branch 'main' into class_templates

Commit:
Update test/Interop/Cxx/templates/Inputs/SwiftClassInstantiationModule.swift

Co-authored-by: Dmitri Gribenko <gribozavr@gmail.com>

hlopko committed:
Address comments

hlopko committed:
Merge branch 'class_templates' of github.com:hlopko/swift into class_templates

hlopko commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -4110,3 +4110,29 @@ clang::FunctionDecl *ClangImporter::instantiateCXXFunctionTemplate(
+  assert(isa<clang::RecordType>(CanonType) &&
+          "type of non-dependent specialization is not a RecordType");
+
+  return dyn_cast_or_null<NominalTypeDecl>(Impl.importDecl(ctsd, Impl.CurrentVersion));
```
Done.

hlopko commented on test/Interop/Cxx/templates/class-template-instantiation-existing-specialization.swift:
```diff
@@ -0,0 +1,19 @@
+  let myInt = IntWrapper(value: 18)
+  var magicInt = MagicWrapper<IntWrapper>(t: myInt)
+  expectEqual(magicInt.getValuePlusArg(12), 30)
+}
```
Done.

hlopko commented on test/Interop/Cxx/templates/class-template-instantiation.swift:
```diff
@@ -0,0 +1,34 @@
+  expectEqual(wrappedMagicNumber.getValuePlusArg(i2), 54)
+}
+
+TemplatesTestSuite.test("existing-specialization") {
```
You're good :) Of course I forgot to delete it when I extracted it to a separete file.

hlopko commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -822,6 +825,45 @@ static Type applyGenericArguments(Type type, TypeResolution resolution,
+          typesOfgenericArgs, templateArguments);
+
+        if (error) {
+            std::string failedTypesStr;
```
Done.

hlopko committed:
Merge main

hlopko committed:
Merge branch 'main' into class_templates

hlopko committed:
Add some header guards

hlopko committed:
Fix bad merge

hlopko committed:
Add simple tests for templates with namespaces

hlopko committed:
WIP

hlopko committed:
WIP

hlopko committed:
Serialize clang template type before instantiation

hlopko committed:
Cleanup

hlopko committed:
Polish

hlopko committed:
Polish

hlopko committed:
Add (failing) test for nested types

hlopko commented:
@swift-ci please test.

hlopko commented:
@swift-ci please test.

hlopko committed:
Merge branch 'main' into class_templates

hlopko committed:
Fix test

hlopko commented:
@swift-ci please test.

hlopko committed:
Merge branch 'main' into class_templates

hlopko commented:
(Failure on mac came from lldb, not related to this PR)

hlopko commented:
@swift-ci please test.

gribozavr commented on include/swift/AST/Decl.h:
```diff
@@ -3401,6 +3401,15 @@ class EnumDecl final : public NominalTypeDecl {
 class StructDecl final : public NominalTypeDecl {
   SourceLoc StructLoc;
+  // We import C++ class template instantiations as non-generic structs
+  // with a name prefixed with `__CxxTemplateInst`. However for proper
+  // serialization we need to have an access to the bound generic type
+  // that would have produced this instantiation. This field contains
+  // that type.
```
"For proper serialization" does not explain the difficultly, only asserts that this is the right solution. Could you add a bit more of an explanation why we need to serialize the bound generic?

Specifically, you should say that when we deserialize a reference to a class template instantiation imported in Swift, we need to have both the Swift struct in the Swift AST and the Clang class template specialization in the Clang AST. Deserialization process will create the Swift struct decl, so we are covered there. However, the corresponding Clang class template specialization is not guaranteed to already exist in any of the Clang modules that we imported so far. Therefore, we need to either instantiate the template during deserialization (the currently implemented solution), or serialize the class template specializations that Swift typechecker asks Clang to create (the ideal long-term solution that's not implemented now).

hlopko replied:
Done, ptal :)

gribozavr commented on test/Interop/Cxx/templates/Inputs/class-template-nested-type-for-swift-module.h:
```diff
@@ -0,0 +1,11 @@
+template <class T> struct Ship {
+    struct Engine {
+        T power;
+    };
```
Indentation is 2 spaces.

hlopko replied:
Done.

gribozavr commented on test/Interop/Cxx/templates/Inputs/class-template-in-namespace-for-swift-module.h:
```diff
@@ -0,0 +1,12 @@
+#define TEST_INTEROP_CXX_TEMPLATES_INPUTS_CLASS_TEMPLATE_IN_NAMESPACE_FOR_SWIFT_MODULE_H
+
+namespace Space {
+    template <class T> struct Ship { T t; };
```
Indentation is 2 spaces.

hlopko replied:
Done.

gribozavr commented on test/Interop/Cxx/templates/class-template-in-namespace-module-interface.swift:
```diff
@@ -5,4 +5,4 @@
 // CHECK:   }
 // CHECK:   typealias Orbiter = Space.__CxxTemplateInstN5Space4ShipIJFvbEEEE
-// CHECK: }
+// CHECK: }
```
Please undo this change.

hlopko replied:
Done.

gribozavr commented on test/Interop/Cxx/templates/class-template-instantiation-existing-specialization.swift:
```diff
@@ -0,0 +1,20 @@
+
+// This test covers the case where Clang has to create specialization eagerly
+// because the code in the header forces it to. Swift has to reuse this
+// specialization, it must not create it's own.
```
```suggestion
// specialization, it must not create its own.
```

Commit:
Update lib/Serialization/Deserialization.cpp

Co-authored-by: Dmitri Gribenko <gribozavr@gmail.com>

Commit:
Update test/Interop/Cxx/templates/class-template-instantiation-existing-specialization.swift

Co-authored-by: Dmitri Gribenko <gribozavr@gmail.com>

hlopko committed:
Addressing comments.

hlopko commented on lib/Serialization/Deserialization.cpp:
```diff
@@ -5353,6 +5354,29 @@ class TypeDeserializer {
+        auto instantiation = clangImporter->instantiateCXXClassTemplate(
+          const_cast<clang::ClassTemplateDecl *>(ctd), templateArguments);
+
+        instantiation->setTemplateInstantiationType(BoundGenericType::get(nominal, parentTy, genericArgs));
```
Done.

hlopko commented on lib/Serialization/Deserialization.cpp:
```diff
@@ -5353,6 +5354,29 @@ class TypeDeserializer {
 
+    if (auto clangDecl = nominal->getClangDecl()) {
+      if (auto ctd = dyn_cast<clang::ClassTemplateDecl>(clangDecl)) {
+        auto clangImporter = static_cast<ClangImporter *>(nominal->getASTContext().getClangModuleLoader());
```
Done.

hlopko commented:
@swift-ci please test.

hlopko commented:
@swift-ci please test macos platform.

hlopko commented:
@swift-ci please test.
